src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+// After this call the
+// reference will be a strong reference because its refcount is >0, and the
+// referenced object is
+// effectively "pinned". Calling this when the refcount is 0 and the object
+// is unavailable
+// results in an error.
+napi_status napi_reference_addref(napi_env e, napi_ref ref, int* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  int count = reference->AddRef();
+
+  if (result != nullptr) {
+    *result = count;
+  }
+
+  return GET_RETURN_STATUS();
+}
+
+// Decrements the reference count, optionally returning the resulting count. If
+// the result is
+// 0 the reference is now weak and the object may be GC'd at any time if there
+// are no other
+// references. Calling this when the refcount is already 0 results in an error.
+napi_status napi_reference_release(napi_env e, napi_ref ref, int* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  int count = reference->Release();
+  if (count < 0) {
+    return napi_set_last_error(napi_generic_failure);
+  }
+
+  if (result != nullptr) {
+    *result = count;
+  }
+
+  return GET_RETURN_STATUS();
+}
+
+// Attempts to get a referenced value. If the reference is weak, the value might
+// no longer be
+// available, in that case the call is still successful but the result is NULL.
+napi_status napi_get_reference_value(napi_env e,
+                                     napi_ref ref,
+                                     napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(ref);
+  CHECK_ARG(result);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  *result = v8impl::JsValueFromV8LocalValue(reference->Get());
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_open_handle_scope(napi_env e, napi_handle_scope* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsHandleScopeFromV8HandleScope(
+      new v8impl::HandleScopeWrapper(v8impl::V8IsolateFromJsEnv(e)));
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_close_handle_scope(napi_env e, napi_handle_scope scope) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(scope);
+
+  delete v8impl::V8HandleScopeFromJsHandleScope(scope);
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_open_escapable_handle_scope(
+    napi_env e,
+    napi_escapable_handle_scope* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsEscapableHandleScopeFromV8EscapableHandleScope(
+      new v8impl::EscapableHandleScopeWrapper(v8impl::V8IsolateFromJsEnv(e)));
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_close_escapable_handle_scope(
+    napi_env e,
+    napi_escapable_handle_scope scope) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(scope);
+
+  delete v8impl::V8EscapableHandleScopeFromJsEscapableHandleScope(scope);
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_escape_handle(napi_env e,
+       
```

Comment:
@addaleax I implemented `napi_instanceof()` this way because I didn't have access to the implementation of the operator. However, if v8 provides access to an implementation that does not require retaining/acquiring the `FunctionTemplate` then I'd be happy to change this API. Additionally, [it looks like](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/hasInstance) `Symbol.hasInstance()` allows users to customize the behaviour of the `instanceof` operator, so we absolutely have to use `Symbol.hasInstance()` when available.